File: LC 329.cpp
class Solution {
    int dir[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
public:
    //BFS
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        vector<vector<int>> degree(n, vector<int>(m, 0));
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                for (int k = 0; k < 4; ++k) {
                    int x = i + dir[k][0], y = j + dir[k][1];
                    if (x >= 0 && x < n && y >= 0 && y < m && matrix[x][y] > matrix[i][j]) {
                        degree[i][j]++;
                    }
                }
            }
        }

        queue<pair<int,int>>q;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (degree[i][j] == 0) {
                    q.push({i,j});
                }
            }
        }

        int ans = 0;
        while (!q.empty()) {
            ++ans;
            for (int t = q.size(); t > 0; --t) {
                int i = q.front().first, j = q.front().second;
                q.pop();
                for (int k = 0; k < 4; ++k) {
                    int x = i + dir[k][0], y = j + dir[k][1];
                    if (x >= 0 && x < n && y >= 0 && y < m && matrix[x][y] < matrix[i][j]) {
                        --degree[x][y];
                        if (degree[x][y] == 0) {
                            q.push({x, y});
                        }
                    }
                }
            }
        }

        return ans;
    }
};

File: LC 329.go
var dir = [][]int{[]int{0,1}, []int{0,-1}, []int{1,0}, []int{-1,0}}
func longestIncreasingPath(matrix [][]int) int {
    n, m := len(matrix), len(matrix[0])
    vis := make([][]int, n)
    for i := range vis {
        vis[i] = make([]int, m)
    }
    var dfs func(i, j int) int 
    dfs = func(i, j int) int {
        if vis[i][j] > 0 {
            return vis[i][j]
        }
        res := 1
        for k := 0; k < 4; k++ {
            x, y := i + dir[k][0], j + dir[k][1]
            if x >= 0 && x < n && y >= 0 && y < m && matrix[x][y] > matrix[i][j] {
                res = max(res, 1 + dfs(x, y))
            }
        }
        vis[i][j] = res
        return res
    }
    ans := 0
    for i := range matrix {
        for j := range matrix[i] {
            if vis[i][j] == 0 {
                ans = max(ans, dfs(i, j))    
            } 
        }
    }
    return ans
}